features/0547-invite-action-protocol/README.md

```diff
@@ -0,0 +1,132 @@
+
+>`https://didcomm.org/invite-action/0.9`
+
+### Key Concepts
+
+This protocol approximates the flow of a classic two-message request-response protocol, except that two responses are possible rather than one, and both responses are optional.
+
+### Roles
+
+The protocol has two roles: `inviter` and `invitee`.
+
+### States
+
+The `inviter` sends a message inviting the `invitee` to take an action. The message may include an advanced [`~please_ack` decorator](../0317-please-ack/README.md) that invites [ACKs](../0317-please-ack/README.md#advanced-features-experimental) to be sent back from `invitee` to `inviter`, either at the time the invitation is accepted or at the time out outcome for the action is known, or both.
+
+The `inviter` state machine looks like this:
+
+[![inviter states](inviter_states.png)](inviter_states.puml)
+
+The `invitee` state machine looks like this:
+
+[![invitee states](invitee_states.png)](invitee_states.puml)
+
+
+### Messages
+
+The first message in the protocol (the only message that's required) is sent from `inviter` to `invitee`. An example might look like this:
+
+```jsonc
+{
+  "@id": "a3079284-9eed-4c76-8d39-a01c797b0e93",
+  "@type": "https://didcomm.org/invite-action/invite",
+  "goal_code": "automotive.inspect.tire",
+  "please_ack": {
+    "on": ["ACCEPT", "OUTCOME"]
+  }
+}
+```
+
+This message suggests that the `invitee` take an action described by the [goal code](../../concepts/0519-goal-codes/README.md) `automotive.inspect.tire`, and asks the `invitee` to send back an ACK when the message is received, and another when the outcome of the action is known. 
+
+#### Adopted Messages
+
+The generic [ACK](../0015-acks/README.md) message, version 1.0, is adopted into this protocol. This means the type of ACKs in this protocol should be:
+
+>`https://didcomm.org/invite-action/%VER/ack`
+
+## Reference
+
+### Messages Details
+
+The only unique field of an `invite` message is `goal_code`. The semantics of goal codes are described in [RFC 0519]. The `~please_ack` decorator is not required, but if present, it matches the semantics described in the [advanced section of RFC 0317](../0317-please-ack/README.md#advanced-features-experimental).
+
+Some general-purpose mechanisms can also enrich this message:
+
+* The `~timing.expires_time` decorator can be used to constrain how long the `inviter` will wait for an ACK.
+* Individual goal codes can specify that they are to be accompanied by specific attachments. Then the generic DIDComm attachment mechanism can be used to pass data to the `invitee`.
+
+
+## Drawbacks
+
+Why should we *not* do this?
+
+## Rationale and alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not
+choosing them?
+- What is the impact of not doing this?
```

Comment:
Have you looked at "Action Menu"?  I believe both this and Action Menu are trying to solve the exact same problem, but Action Menu does it in a more general way, allowing the user to see a menu of things they can do vs. just one, and allowing the invitee to include data as part of the request to act.  Action Menu also allows the invitee to say "Hey, send me your menu so I can see what services you offer".

I think at least it should be included in the list of alternatives.

Reply:
Yes, I have. As a matter of fact, I re-read its RFC before I started on this. I didn't see it as an exact equivalent, because it  goes the opposite direction (agent willing to act must proactively publish a menu, instead of passively receiving an invitation). But I agree that it should be included in the list of alternatives.

Reply:
Sounds like we didn't explain it well.  It very much was intended to be either side starting the conversation, the "service-provider" offering a menu of options (could be just one as is the case here), and the "service-user" selecting.

We definitely need to get goal-code in into Action Menu, but I would be really interested in trying to merge these and make sure that we are covering the necessary cases.  I think there is too much in common with these approaches.

Reply:
I've added a section to the comparison table talking about Action Menu. I identified the things that we might have to resolve to collapse the two.

Reply:
These are the aspects of Action Menu that stand out to me:

* allows one Agent to present a set of heirarchical menus and actions to another user-facing Agent in a human friendly way
* requires an active DIDComm connection before it can proceed. 
* the responder presents a list of actions which can be initiated by the requester. Actions are contained within a menu structure.
* can be initiated by either the requester asking for the root menu or the responder sending an unsolicited root menu.

It sounds a lot like RFC0031 "Discover Features Protocol" but a little bit more oriented towards enabling a human-friendly display of the actions. It also sounds a lot like a "service provider".

It makes sense for Action Menu to require a DIDComm connection between agents because, just like in RFC0031, an agent may wish to limit service exposure to un-trusted parties. An agent could theoretically send some sort of out-of-band menu request with a goal code and receive a tailored action menu in response. The agent would then need to take their pick.

It was hard for me to grok this inversion because normally human beings and software with logic just ask one another what to do with the expectation that the other party knows what they must do. The expectation is usually derived from the requester's context. If the requester now has to query a navigable menu received in response to the original request and then point the responder to the _exact thing they'd like to be performed_ then that is adding a bit of overhead to the exchange.

Coming back to Daniel's auction example: Alice is auctioning a Rembrandt and would like to invite Bob to bid for it. She needs to send a message to Bob telling him the action, the subject, and the location. Bob either understands the message or he doesn't. He could reply with an Action Menu if he doesn't, but now Alice needs to understand Bob's action menu. 

And I just realized something: the RFC proposed in this PR includes action (goal code) but does not include subject nor location. We're missing something here...

Reply:
A couple of thoughts on @llorllale comments:

- Action Menu and Discover Features are not the same as Discover Features returns a list of protocols supported, while Action Menu is a list "business things you can do at this time".  It's intended to be at a much higher layer.  The fact that an agent supports "Issue credential" tells me nothing about what credentials are offered in the current context.

- Action menus are not hierarchical, although a hierarchy can be constructed via a sequence of interactions.  The telephone IVR system is the target -- at any point, the user gets a menu of things that they can do, including some that lead to other menus, others that lead to talking to a person for a specific reason, others that just send you to the operator.

The use of Action Menu will be needed when an Agent can be used to execute an entire business transaction vs. the current model we are seeing where the transaction is happening in a browser, and the agent is being used to provide a proof or receive a credential as a side-effect of the transaction.  If you could use your agent to contact your bank and execute one of a set of transactions (check balance, transfer funds, open bank, get proof of balance), then the richer features of Action Menu become necessary.

Coming back to Daniel's auction, Alice could be auctioning a Rembrandt and present a menu of actions that Bob can take -- request proof of authenticity, see picture of painting, make a bid, request bid history, exit the bidding.  And since, Alice is running an Auction house, Bob may start at a top menu allowing searching current auctions, see the painting of the day, pay invoice, talk to Alice.

Reply:
@swcurran 

That all sounds good. The Action Menu RFC should be updated to enable all these use cases.

Reply:
I have begun a PR that would remove this RFC and update the Action Menu RFC. I'll try to finish that up.